src/providers/wfs/qgsoapifprovider.cpp

```diff
@@ -337,7 +337,14 @@ const QString &QgsOapifProvider::clientSideFilterExpression() const
 
 void QgsOapifProvider::handlePostCloneOperations( QgsVectorDataProvider *source )
 {
-  mShared = qobject_cast<QgsOapifProvider *>( source )  ->mShared;
+  disconnect( mShared.get(), &QgsOapifSharedData::raiseError, this, &QgsOapifProvider::pushErrorSlot );
+  disconnect( mShared.get(), &QgsOapifSharedData::extentUpdated, this, &QgsOapifProvider::fullExtentCalculated );
+
+  // We must not change the subset string of the shared data used in another iterator/data provider ...
+  mShared.reset( qobject_cast<QgsOapifProvider *>( source )->mShared->clone() );
```

Comment:
Can we stick to doing that just after set subset string or not?

If it's done in here we run some risks again:
- we have the subset string in a shared data that is changed while iterators are running on it, this looks risky to me
- feature ids will be changed on cloned layers which was the original issue related to the label export algorithm having wrong references to original features

can we move this only to post-subset-string change?

Reply:
> can we move this only to post-subset-string change?

I actually tried this. But there was one unit test that breaks in [providertestbase.py](https://github.com/qgis/QGIS/pull/48998/files#diff-e112d0d5d27c1dfff4ecd30f58bb07a19c07051a8f2c71ae4f9dbadf85146f77) at the place where I left a comment. This test is related to feature id stability. The fact that this test passes with the current code is mostly a miracle (actually when running this setSubsetString() test in isolation and not the whole test_provider_wfs.py file, it breaks, probably because it depends on other tests to have previously loaded the whole layer in the cache). So maybe we should do just a hack here and not try this test for the WFS provider

Regarding "feature ids will be changed on cloned layers", there might be an issue in using the same SharedData object by 2 layers. If the 2 layers are rendered in separate threads, the SharedData would be used (and the objects it contains, in particular the FID<->gml:id cache) concurrently, and bad things could happen. So it seems safer to me to clone the SharedData as a post processing step of layer cloning. Ideally we'd want to duplicate the FID<-->gml:id cache (which resides in a SQLite db), but that isn't not implemented currently.

Reply:
> > can we move this only to post-subset-string change?
> 
> I actually tried this. But there was one unit test that breaks in [providertestbase.py](https://github.com/qgis/QGIS/pull/48998/files#diff-e112d0d5d27c1dfff4ecd30f58bb07a19c07051a8f2c71ae4f9dbadf85146f77) at the place where I left a comment. This test is related to feature id stability. The fact that this test passes with the current code is mostly a miracle (actually when running this setSubsetString() test in isolation and not the whole test_provider_wfs.py file, it breaks, probably because it depends on other tests to have previously loaded the whole layer in the cache). So maybe we should do just a hack here and not try this test for the WFS provider

Skipping this test sounds reasonable with this explanation.

> Regarding "feature ids will be changed on cloned layers", there might be an issue in using the same SharedData object by 2 layers. If the 2 layers are rendered in separate threads, the SharedData would be used (and the objects it contains, in particular the FID<->gml:id cache) concurrently, and bad things could happen. So it seems safer to me to clone the SharedData as a post processing step of layer cloning. Ideally we'd want to duplicate the FID<-->gml:id cache (which resides in a SQLite db), but that isn't not implemented currently.

One single layer can already now have iterators running in different threads, so concurrency shouldn't be something new to handle, unless I am missing some of the magic of the WFS provider.

Reply:
> can we move this only to post-subset-string change?

I've implemented that in the latest commit

> One single layer can already now have iterators running in different threads, so concurrency shouldn't be something new to handle

hum, there might be potential troubles then as I doubt the code is ready for that. (and that would be probably hard to achieve)

Reply:
Ah, sorry I have missed this comment.
In this case I would say let's merge this, it only improves the situation.
In parallel, should we file a ticket to review multithreading here for an upcoming bugfix sprint?

Reply:
> In parallel, should we file a ticket to review multithreading here for an upcoming bugfix sprint?

done in https://github.com/qgis/QGIS/issues/49181